### PR TITLE
Datastore Query 

### PR DESCRIPTION
**Pull request by BenMob**
Hello team. Please review! 

Summary
- Handles POST requests from the search page to Datastore.
- Sets up a Query that returns a JSON response of students filtered by subject at  ```/data```.

Thanks :)
Benjamin Ombeni

### CODE REVIEW

File: studyBuddy/src/main/java/com/google/sps/models/Student.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+package com.google.sps.models;
+
+public final class Student{
+    private String firstName;
+    private String lastName;
+    private String school;
+    private String subject;
+    private long timestamp;
+
+    public Student(String firstName, String lastName, String school, String subject, long timestamp){
```
Comment (sgh005) [inline studyBuddy/src/main/java/com/google/sps/models/Student.java:12]: looks gr8, just wanted to clarify that the login info would fill most of this out and so when we search we essentially just take the subject to fill the rest? but going to approve (:

Comment (BenMob) [reply studyBuddy/src/main/java/com/google/sps/models/Student.java:12]: Hi @sgh005 

If I understand you are saying that the student's login info would fill in most of the attributes in the student class model. 

Yes. The Authentication feature will fill in the following:

- ```email``` (which I forgot to add to the class model but I will do it lol)
- ```last name``` (will be part of the email I think)
- ```school``` (will be part of the email for sure)

If we want to collect the first name as well, we may need an additional **Sign Up Page** where we prompt for the ```first name``` (or simply a username/nickname) or something of that sort because the authentication library only gives us access to the user's **email address**. 

Any suggestions?

Thanks for the review :) 

Ben

Comment (sgh005) [reply studyBuddy/src/main/java/com/google/sps/models/Student.java:12]: Makes sense! No yeah that works I was just making sure I understood our implementation. Also for whatever reason I am unable to approve the request as it needs Michaels approval. 
-Steph

File: studyBuddy/src/main/java/com/google/sps/models/Student.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
+// This is a Student class model 
+
+package com.google.sps.models;
+
+public final class Student{
```
Comment (mpasetes17) [inline studyBuddy/src/main/java/com/google/sps/models/Student.java:5]: I think we want a field for email too, right?
It looks like you just need to pass it from createStudentFromEntity.

File: studyBuddy/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 3]
```diff
@@ -19,14 +19,133 @@
...
+********************/
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+    // TODO Integrate UserService
+    final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    
+    // Query Datastore
+    Filter subject_filter = new FilterPredicate("subject", FilterOperator.EQUAL, this.subject);
```
Comment (mpasetes17) [inline studyBuddy/src/main/java/com/google/sps/servlets/DataServlet.java:51]: I believe this is only filtering by subject, which will work. But I don't know if we also wanted to make sure that the schools matched too.

File: studyBuddy/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 4]
```diff
@@ -19,14 +19,133 @@
...
+    // Scan results and add the in the students ArrayList
+    for (Entity entity: results.asIterable()){
+        Student current_student = createStudentFromEntity(entity);
+        students.add(current_student);
+    }
+
+    // Return a JSON response of the results
+    response.setContentType("application/json;");
+    response.getWriter().println(convertToJson(students));
+  }
```
Comment (mpasetes17) [inline studyBuddy/src/main/java/com/google/sps/servlets/DataServlet.java]: I'm not sure how this is going to work. From our design meeting, I thought the plan was to create a single POST request that searches for a match and sends the person's info to the datastore, but here it looks like that is split across a GET and POST which I believe will create a race condition.

One solution might be to move the functionality of the GET request to a private function and call it from inside the POST request.

File: studyBuddy/src/main/java/com/google/sps/models/Student.java (hunk 2 above)
Comment (mpasetes17) [reply studyBuddy/src/main/java/com/google/sps/models/Student.java:5]: My bad, I just saw the other comments.

File: studyBuddy/src/main/java/com/google/sps/models/Student.java (hunk 1 above)
Comment (mpasetes17) [reply studyBuddy/src/main/java/com/google/sps/models/Student.java:12]: Having a first and last name would be a good feature, which we might be able to add once we have the idea of a user profile. But for now, I think we will want to just use the email address as the identifier because it will be a unique to each user and doesn't require us setting up any additional inputs.

As for the whole last name thing, I don't think it is standard to use last names as emails (my school doesn't, because there are too many students with the same last name).

**Comment by mpasetes17** [comment]
Awesome, it looks like we have a lot of the core functionality just about ready. 

I'm not super familiar with how all of the filtering on the data queries works. Were you able to do any unit testing, or do we want to wait until we have more of the linkages ready to go?

File: studyBuddy/src/main/java/com/google/sps/models/Student.java (hunk 1 above)
Comment (BenMob) [reply studyBuddy/src/main/java/com/google/sps/models/Student.java:12]: Hi @mpasetes17 

Understood! 

- I built the Entity based on the user sample that is on Search Page.  
  If you look closely I hardcoded the ```first name: John , last name: Doe``` 
 
- Give me a suggestion here: Should I get rid of the first and last name properties for now or should I just leave them there but hardcoded to illustrate a potential integration of the feature?  
  
- Thanks for letting me know that last names in school emails are not standard. I will keep that in mind. 


Ben

File: studyBuddy/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 3 above)
Comment (BenMob) [reply studyBuddy/src/main/java/com/google/sps/servlets/DataServlet.java:51]: Yes. We would want to filter by School as well. I will add that once I integrate the user authentication feature because that is where we will grab information about the school.
For now, it is just filtering based on the information coming from the search page i.e **subject**

Thanks.

File: studyBuddy/src/main/java/com/google/sps/models/Student.java (hunk 1 above)
Comment (mpasetes17) [reply studyBuddy/src/main/java/com/google/sps/models/Student.java:12]: I don't think there is any harm in leaving them in there, especially since this will be something that we would like to implement fairly soon after we get the MVP. I would say go ahead and leave the field, but hard code the value so that it will be easily distinguishable as not valid data about the entry. I don't know how exactly we want to lay out the data store table, but this could be as simple as setting both to NULL until we are able to implement that feature.

File: studyBuddy/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 3 above)
Comment (mpasetes17) [reply studyBuddy/src/main/java/com/google/sps/servlets/DataServlet.java:51]: Sounds good.

File: studyBuddy/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 4 above)
Comment (BenMob) [reply studyBuddy/src/main/java/com/google/sps/servlets/DataServlet.java]: Thanks for the feedback. I will address this.

@mpasetes17 @sgh005 
Quick question for clarification: Among students who match  the ```subject``` and ```school``` criteria requested, are we only sending back 

- A single student. 
- All students.
- Or maybe a Top N number of students based on ```timestamp```.

For now, it is returning **all**. 

Ben

Comment (mpasetes17) [reply studyBuddy/src/main/java/com/google/sps/servlets/DataServlet.java]: I think returning all is fine. If it becomes an issue, I don't think it will take much to add a limit.

**Comment by BenMob** [comment]
@mpasetes17 I did not do any Unit Testing as I am not very familiar with that part yet. However, I will review Unit Testing a little more sometimes this week and do it.

I tested it manually by doing different POST requests from the client and analyzing each result. Not the best way to test though.

Ben

**Comment by mpasetes17** [comment]
@BenMob, manual testing is good enough. Obviously, having written ones is more convenient, but given the time restraints, I don't think it would be the best use of your time the moment.

If you remember the scenarios that you ran through, it may be a good idea to write some of them out in this thread, just so that there is documentation on what this code is supposed to do exactly, particularly if there are any edge cases. I just want to make sure that the things that get added later don't cause problems with this, especially since it is the core of the app.

**Comment by BenMob** [comment]
Quick Test Demo:

- **This is how datastore looks after a few post requests:**

Screen Shot 2020-03-23 at 7 31 56 AM

- **Requesting match with all **math** students:**

Screen Shot 2020-03-23 at 7 32 52 AM

- **Results: All math students**
Screen Shot 2020-03-23 at 7 33 41 AM

Note: Ignore the first and last names being John Doe at certain places. Those were posted before I changed their values to NULL. 

One more thing: The person who just submitted a request is also being included in the results for now. I am still going to restrict that in the query as soon as I can.
 
Alright that is it.

Ben

**Review by mpasetes17 — APPROVED** [review]
Sorry for the late review.
Looks good!
